torch/distributed/fsdp/optim_utils.py

```diff
@@ -192,11 +221,95 @@ def _unflatten_communicated_optim_state(
+        List[Dict[str, Any]], Iterable[torch.nn.Parameter],
+    ]] = None,
+) -> Tuple[Dict[str, Any], Set[int]]:
+    """
+    Args:
+        shard_state (bool): Whether to shard flattened positive-dimension
+            tensor state; if ``False``, then the full flattened tensor is
+            kept in the returned :class:`dict.
+
+    Returns:
+        Tuple[Dict[str, Any], Set[int]]: The flattened optimizer state dict
+            and a set of the parameter IDs corresponding to FSDP parameters.
+    """
+    full_osd = full_optim_state_dict  # alias
+    if "state" not in full_osd or "param_groups" not in full_osd:
+        raise ValueError(
+            "`full_optim_state_dict` must have the keys \"state\" and "
+            "\"param_groups\" to be a valid optimizer state dict"
+        )
+
+    flat_param_id_to_param = _get_param_id_to_param(model, optim_input)
+    flat_param_to_fsdp_module = _get_flat_param_to_fsdp_module(model)
+    param_to_unflat_param_names = FSDP._get_param_to_unflat_param_names(model)
+
+    # Handle the "state" part of the optimizer state dict
+    flat_osd_state: Dict[int, Any] = {}
+    full_osd_state = full_osd["state"]
+    unflat_param_names_to_flat_param_id: Dict[str, int] = {}
+    fsdp_flat_param_ids = set()  # save which IDs are for FSDP parameters
+    for flat_param_id, param in enumerate(flat_param_id_to_param):  # type: ignore[assignment]
+        assert param in param_to_unflat_param_names, \
+            "Check the `param_to_unflat_params` construction\n" \
+            f"param: {param}"
+        unflat_param_names = param_to_unflat_param_names[param]
+        # For FSDP parameters, we need to flatten
+        if isinstance(param, FlatParameter):
+            assert param in flat_param_to_fsdp_module, \
+                "Check the `flat_param_to_fsdp_module` mapping " \
+                f"construction\nparam={param}"
+            unflat_param_names = param_to_unflat_param_names[param]
+            fsdp_module = flat_param_to_fsdp_module[param]
+            flat_state = _flatten_optim_state(
+                full_osd_state, unflat_param_names, fsdp_module, param,
+                shard_state,
+            )
+            flat_osd_state[flat_param_id] = flat_state
+            for unflat_param_name in unflat_param_names:
+                unflat_param_names_to_flat_param_id[unflat_param_name] = flat_param_id
+            fsdp_flat_param_ids.add(flat_param_id)
+        # For parameters from non-FSDP modules, we do not need to flatten
+        else:
+            assert len(unflat_param_names) == 1
+            unflat_param_name = unflat_param_names[0]
+            # Remap from unflattened to flattened parameter ID -- do not
+            # deepcopy to avoid unnecessarily duplicating tensor storage
+            flat_osd_state[flat_param_id] = \
+                copy.copy(full_osd_state[unflat_param_name])
+            unflat_param_names_to_flat_param_id[unflat_param_name] = flat_param_id
+
```

Comment:
No need since the `fsdp_flat_param_ids` is meant specifically to identify which IDs correspond to `FSDP` flattened parameters. This is to know which tensor state needs to be sharded by rank 0 and broadcast shard by shard as opposed to not sharded and broadcast as is.